contracts/token/ERC20/SafeERC20.sol

```diff
@@ -5,36 +5,78 @@ import "../../math/SafeMath.sol";
  */
 library SafeERC20 {
     using SafeMath for uint256;
 
     function safeTransfer(IERC20 token, address to, uint256 value) internal {
-        require(token.transfer(to, value));
+        callAndAssertSuccess(token, abi.encodeWithSignature("transfer(address,uint256)", to, value));
     }
 
     function safeTransferFrom(IERC20 token, address from, address to, uint256 value) internal {
-        require(token.transferFrom(from, to, value));
+        callAndAssertSuccess(token, abi.encodeWithSignature("transferFrom(address,address,uint256)", from, to, value));
     }
 
     function safeApprove(IERC20 token, address spender, uint256 value) internal {
         // safeApprove should only be called when setting an initial allowance,
         // or when resetting it to zero. To increase and decrease it, use
         // 'safeIncreaseAllowance' and 'safeDecreaseAllowance'
         require((value == 0) || (token.allowance(address(this), spender) == 0));
-        require(token.approve(spender, value));
+        callAndAssertSuccess(token, abi.encodeWithSignature("approve(address,uint256)", spender, value));
     }
 
     function safeIncreaseAllowance(IERC20 token, address spender, uint256 value) internal {
         uint256 newAllowance = token.allowance(address(this), spender).add(value);
-        require(token.approve(spender, newAllowance));
+        callAndAssertSuccess(token, abi.encodeWithSignature("approve(address,uint256)", spender, newAllowance));
     }
 
     function safeDecreaseAllowance(IERC20 token, address spender, uint256 value) internal {
         uint256 newAllowance = token.allowance(address(this), spender).sub(value);
-        require(token.approve(spender, newAllowance));
+        callAndAssertSuccess(token, abi.encodeWithSignature("approve(address,uint256)", spender, newAllowance));
+    }
+
+    /**
+     * @dev Performs a function call on a token, asserting its success, which is defined by a) the call succeeding,
+     * and b) the return data being either empty or true.
+     * @param token The token targeted byt he call.
+     * @param data The call data (encoded using abi.encode or one of its variants).
+     */
+    function callAndAssertSuccess(IERC20 token, bytes memory data) private {
+        // We need to perform a low level call here, to bypass Solidity's return data size checking mechanism, since
+        // we're implementing it ourselves.
+
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success,) = address(token).call(data);
+        require(success);
+        
+        require(getBooleanReturnValue());
+    }
+
+    /**
+     * @dev Parses the return value of the last contract call and attempts to return a boolean value out of it.
+     */
+    function getBooleanReturnValue() private pure returns (bool result) {
+        // We need to use inline assembly here, since it is the only way we can utilize the returndatasize opcode.
```

Comment:
This is true for the _opcode_, but we could be using the `data` byte array returned by `call` and reading `data.length`. What do you think?

I don't feel so comfortable having this function use the `returndata` opcodes because they're like reading from global variables.

Reply:
`data.length` sure, but I'm not sure if there's a convenient way to check for a boolean value stored inside the `bytes` array.

Reply:
Maybe it's a good time to introduce that bytes conversion library. :eyes: